File: Recursion/howSum.js
let calls = 0

const howSum = (targetSum, numbers, memo={}) => {
  calls += 1
  if(targetSum in memo) return memo[targetSum]

  if(targetSum === 0) return []

  if(targetSum < 0) return null

  for(let num of numbers){

    const remainder = targetSum - num
    
    const result = howSum(remainder, numbers, memo)
    
    if(result !== null){
      memo[targetSum] = result

      return [...result, num]
    }

  }
  
  memo[targetSum] = null
  return null

}

console.log(howSum(7, [5, 3, 4, 7]), calls)
calls = 0
console.log(howSum(300, [7, 14]), calls)

File: Recursion/howSum.py
def howSum(targetSum, array, memo={}):
  #print(memo)
  if targetSum in memo:
    return memo[targetSum]

  if targetSum == 0:
    return []

  if targetSum < 0: 
    return None

  for num in array:
    remainder = targetSum - num
    result = howSum( remainder, array )

    if result is not None:
      memo[targetSum] = result
      return result + [num]

  memo[targetSum] = None
  return None
print("Answer: ", howSum(59, [14, 4, 6, 5, 2]))